docs/adr/adr-11-blocksync-overhaul.md

```diff
@@ -0,0 +1,329 @@
+  carIdx: index.NewFSRepo(basepath + "/index")
+  mounts: mounts,
+    }   
+}
+
+```
+
+##### `share.EDSStore.Put`
+
+To write an entire EDS `Put` method is introduced. Internally it
+
+- Opens a file under `storepath/DataRoot` path
+- Serializes the EDS into the file via `share.WriteEDS`
+- Wraps it with `DAGStore`'s [FileMount](https://github.com/filecoin-project/dagstore/blob/master/mount/file.go#L10)
+- Converts `DataRoot` into the [`shard.Key`](https://github.com/filecoin-project/dagstore/blob/master/shard/key.go#L12)
+- Registers the `Mount` as a `Shard` on the `DAGStore`
+  - This register the `Mount` in [`mount.Registry`](https://github.com/filecoin-project/dagstore/blob/master/mount/registry.go#L22)
+passed to `DAGStore`.
+
+NOTE: Registering on the DAGStore automatically populates top-level index with shares/proofs accessible from stored EDS
+so this is out of scope of the document.
+
+```go
+// Put stores the given data square with DataRoot as key.
+//
+// The square is verified on the Exchange level and Put only stores the square trusting it.
+// The resulting file stores all the shares and NMT Merkle Proofs of the EDS.
+// Additionally, the file gets indexed s.t. Store.Blockstore can access them. 
+func (s *Store) Put(context.Context, DataRoot, *rsmt2d.ExtendedDataSquare) error
+```
+
+##### `share.EDSStore.GetCAR`
+
+To read an EDS as a byte stream `GetCAR` method is introduced. Internally it
+
+- Converts `DataRoot` into the [`shard.Key`](<https://github.com/filecoin-project/dagstore/blob/master/shard/key.go#L12>
+- Gets Mount by `shard.Key` from [`mount.Registry`](https://github.com/filecoin-project/dagstore/blob/master/mount/registry.go#L22)
+- Returns `io.ReadeCloser` from [`Mount.Fetch`](https://github.com/filecoin-project/dagstore/blob/master/mount/mount.go#L71)
+
+NOTE: It might be necessary to acquire EDS mount via `DAGStore` keeping `ShardAccessor`
+and closing it when operation is done. This has to be confirmed.
+
+```go
+// GetCAR takes DataRoot and returns a buffered reader to respective EDS serialized as CARv1 file.
+// 
+// The Reader strictly reads the first quadrant(1/4) of EDS omitting all the NMT Merkle proofs.
+// Integrity of the store data is not verified. 
+// 
+// Caller must Close returned reader after reading.
+func (s *Store) GetCAR(context.Context, DataRoot) (io.ReadCloser, error)
+```
+
+##### `share.EDSStore.Blockstore`
+
+`Blockstore` method return a `Blockstore` interface implementation instance, providing random access over share and NMT
+Merkle proof in every stored EDS. It is required for FNs/BNs to serve DAS requests over the Bitswap and for reading data
+by namespace. // TODO Link to the section
+
+There is a [frozen/un-merged implementation](https://github.com/filecoin-project/dagstore/pull/116) of `Blockstore`
+over `DAGStore` and CARv2 indexes.
+
+NOTE: We can either use it(and finish if something is missing for our case) or implement custom optimized for our needs.
```

Comment:
It does not really fit our case exactly, I think we should rather implement a custom one (and it will probably look pretty similar to what i have in the draft PR already anyways, so there is only very little effort needed for this)

Reply:
the rationale is in the go doc of my experimentations i'll post it when i have PC access again

And: we should have absolutely 0 bias towards keeping what I've already done, if we throw it all away I still gained very very much by doing it! The hope is of course that it could uncover some details that would help the ADR or downstream impl, but that's really it

Reply:
Mostly, the custom implementation is simpler (doesn't take things we don't need) and doesn't throw on Put/PutMany which would break the exchange otherwise (blockservices call put on the blockstore)

```
// EDSStore implements the blockstore interface on a DAGStore.
// The lru cache approach is heavily inspired by the open PR filecoin-project/dagstore/116.
// The main differences to the implementation here are that we do not support multiple shards per key,
// call GetSize directly on the underlying RO blockstore, and do not throw errors on Put/PutMany.
```

Reply:
https://github.com/celestiaorg/celestia-node/pull/1037#discussion_r971023278